### PR TITLE
raftstore-v2: reduce raft log size

### PR DESCRIPTION
**Pull request by tabokie**
### What is changed and how it works?

Issue Number: Ref #12842 

What's Changed:

- manual flush regions that need force compact
- default flush oldest first
- increase recovery threads to min(1.5*cpu, 16)
- ensure purge-threshold >= write-buffer-limit * 2

```commit-message
Reduce raft log size of raftstore-v2
```

### Related changes

https://github.com/tikv/rust-rocksdb/pull/742
https://github.com/tikv/rocksdb/pull/337

### Check List 

Tests 

- Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by tabokie** [comment]
image

**Comment by dbsid** [comment]
after 10 hours stability test, we don't see rewrite flow, but the number of raftlog file grows to 3801 and raft-engine folder take 477GB.

image


```
[root@tc-tikv-0 data]# du -sh *
0       LOCK
12K     import
24K     last_tikv.toml
477G    raft-engine
0       raftdb.info
301M    rocksdb-2023-05-20T00-44-23.067.info
301M    rocksdb-2023-05-20T03-08-24.029.info
301M    rocksdb-2023-05-20T05-15-40.360.info
166M    rocksdb.info
4.0K    tablet_snap
430G    tablets
[root@tc-tikv-0 data]# cd raft-engine
[root@tc-tikv-0 raft-engine]# ls -l *raftlog | wc -l
3801
[root@tc-tikv-0 raft-engine]# ls -l *rewrite | wc -l
17
[root@tc-tikv-0 raft-engine]# cd /
[root@tc-tikv-0 /]# ./tikv-server --version
TiKV 
Release Version:   7.2.0-alpha
Edition:           Community
Git Commit Hash:   cb371fce8f2da03fcde0232c9f0881b2bb367bdc
Git Commit Branch: heads/refs/tags/v7.2.0
UTC Build Time:    2023-05-19 06:43:57
Rust Version:      rustc 1.67.0-nightly (96ddd32c4 2022-11-14)
Enable Features:   pprof-fp jemalloc mem-profiling portable sse test-engine-kv-rocksdb test-engine-raft-raft-engine cloud-aws cloud-gcp cloud-azure
Profile:           dist_release
```

```
[raft-engine]
  enable-log-recycle = true
  purge-threshold = "50GB"

[raftstore]
  store-io-pool-size = 2

[rocksdb]
  rate-bytes-per-sec = "0B"
  [rocksdb.defaultcf]
    compression-per-level = ["zstd", "zstd", "zstd", "zstd", "zstd", "zstd", "zstd"]
    use-bloom-filter = false
  [rocksdb.lockcf]
    use-bloom-filter = false
  [rocksdb.writecf]
    use-bloom-filter = false

[storage]
  engine = "partitioned-raft-kv"
  reserve-space = "0MB"
  [storage.io-rate-limit]
    level-zero-compaction-priority = "low"
    max-bytes-per-sec = "600MiB"
```

**Comment by dbsid** [comment]
After the latest fix, the disk usage of raft-engine is stable at 52GB, the memory usage is reduced by an order of magnitude.

image

```
[root@tc-tikv-0 raft-engine]# du -sh .
52G    
```

File: components/raftstore-v2/src/batch/store.rs [hunk 1]
```diff
@@ -616,12 +618,29 @@ impl<EK: KvEngine, ER: RaftEngine> StoreSystem<EK, ER> {
...
+            let registry = tablet_registry.clone();
             worker.spawn_interval_task(cfg.value().raft_engine_purge_interval.0, move || {
+                let limiter = Limiter::new(MAX_MANUAL_FLUSH_RATE);
                 let _guard = WithIoType::new(IoType::RewriteLog);
                 match raft_clone.manual_purge() {
-                    Ok(regions) => {
+                    Ok(mut regions) => {
+                        warn!(logger, "flushing oldest cf of regions {regions:?}");
+                        // Refresh the list at least every 60s.
+                        regions.truncate((MAX_MANUAL_FLUSH_RATE * 60.0) as usize);
```
Comment (LykxSassinator) [inline components/raftstore-v2/src/batch/store.rs:629]: Here, `MAX_MANUAL_FLUSH_RATE` == 0.3 * 60.0 => 18 as default, an empirical value proved appropriate for `raftstore-v2`?

> Refresh the list at least every 60s.

And what's the meaning of this annotation? Might not clear for this.

Comment (tabokie) [reply components/raftstore-v2/src/batch/store.rs:629]: 0.3 is the max speed, 0.3*60 is the maximum memtables we can flush within 60s.

Comment (LykxSassinator) [reply components/raftstore-v2/src/batch/store.rs:629]: got.

**Review by SpadeA-Tang — APPROVED** [review]

File: components/engine_rocks/src/misc.rs [hunk 2]
```diff
@@ -138,14 +138,52 @@ impl MiscExt for RocksEngine {
...
+        }
+        if let Some((handle, time)) = handles
+            .into_iter()
+            .filter_map(|handle| {
+                self.as_inner()
+                    .get_approximate_active_memtable_stats_cf(handle)
+                    .map(|(_, time)| (handle, time))
+            })
+            .min_by(|(_, a), (_, b)| a.cmp(b))
+            && age_threshold.map_or(true, |threshold| time <= threshold)
```
Comment (tonyxuqqi) [inline components/engine_rocks/src/misc.rs:174]: |threshold| time <= threshold

> Should it be > per comment said.

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**